audio/out/ao_aaudio.c

```diff
@@ -0,0 +1,503 @@
+        if (af_fmt_to_bytes(ao->format) > 2 && p->device_api >= 31) {
+            ao->format = AF_FORMAT_S32;
+            format = AAUDIO_FORMAT_PCM_I32;
+        } else {
+            ao->format = AF_FORMAT_S16;
+            format = AAUDIO_FORMAT_PCM_I16;
+        }
+    } else {
+        ao->format = AF_FORMAT_S16;
+        format = AAUDIO_FORMAT_PCM_I16;
+    }
+
+    p->AAudioStreamBuilder_setDeviceId(builder, p->device_id);
+    p->AAudioStreamBuilder_setDirection(builder, AAUDIO_DIRECTION_OUTPUT);
+    p->AAudioStreamBuilder_setSharingMode(builder,
+                                          (ao->init_flags & AO_INIT_EXCLUSIVE)
+                                              ? AAUDIO_SHARING_MODE_EXCLUSIVE
+                                              : AAUDIO_SHARING_MODE_SHARED);
+    p->AAudioStreamBuilder_setFormat(builder, format);
+    p->AAudioStreamBuilder_setSampleRate(builder, ao->samplerate);
+    p->AAudioStreamBuilder_setErrorCallback(builder, error_callback, ao);
+    p->AAudioStreamBuilder_setBufferCapacityInFrames(builder,
+                                                     p->buffer_capacity);
+    p->AAudioStreamBuilder_setPerformanceMode(builder, p->performance_mode);
+    p->AAudioStreamBuilder_setDataCallback(builder, data_callback, ao);
+
+    if (p->device_api >= 28) {
+        p->AAudioStreamBuilder_setContentType(builder,
+                                              (ao->init_flags &
+                                               AO_INIT_MEDIA_ROLE_MUSIC)
+                                                  ? AAUDIO_CONTENT_TYPE_MUSIC
+                                                  : AAUDIO_CONTENT_TYPE_MOVIE);
+        p->AAudioStreamBuilder_setUsage(builder, AAUDIO_USAGE_MEDIA);
+        p->AAudioStreamBuilder_setSessionId(builder, p->session_id);
+    }
+
+    if (p->device_api >= 32) {
+        uint8_t i = 0;
+        struct mp_chmap channels = ao->channels;
+
+        do {
+            struct mp_chmap_sel sel = {0, .tmp = p};
+            aaudio_channel_mask_t channel_mask = AAUDIO_CHANNEL_INVALID;
+
+            for (int j = 1; j < aaudio_max_chnums[i]; j++)
+                mp_chmap_sel_add_map(&sel, &aaudio_chmaps[j]);
+
+            if (!ao_chmap_sel_adjust(ao, &sel, &channels)) {
+                MP_ERR(ao, "Failed to find channel map\n");
+                goto err;
+            }
+
+            for (uint8_t j = 0; j < aaudio_max_chnums[i]; j++) {
+                if (mp_chmap_equals(&aaudio_chmaps[j], &channels)) {
+                    channel_mask = aaudio_masks[j];
+                    break;
+                }
+            }
+            assert(channel_mask != AAUDIO_CHANNEL_INVALID);
+
+            p->AAudioStreamBuilder_setChannelMask(builder, channel_mask);
+        } while (i++ < MP_ARRAY_SIZE(aaudio_max_chnums) &&
+                 (result = p->AAudioStreamBuilder_openStream(builder, &p->stream)) ==
+                     AAUDIO_ERROR_OUT_OF_RANGE);
```

Comment:
please don't have the success/error condition in the while condition, that's confusing